CorsixTH/Lua/game_ui.lua

```diff
@@ -117,6 +117,24 @@ function GameUI:makeVisibleDiamond(scr_w, scr_h)
   }
 end
 
+--! Calculate the minimum valid zoom value
+--!
+--! Zooming out too much would cause negative width/height to be returned from
+--! makeVisibleDiamond. This function calculates the minimum zoom_factor that
+--! would be allowed.
+function GameUI:calculateMinimumZoom()
```

Comment:
From what I can tell `adjustZoom` relies on `setZoom` failing to limit zooming out (it does provide the maximum zoom as a 'virtual width' of 200). `setZoom` fails when `makeVisibleDiamond` returns a negative width or height. The logic in `calculateMinimumZoom` simply working out when the width/height would be 0.

Unfortunately having testing this a bit more I've found that it is still possible to trigger the behaviour where scrolling/zooming doesn't work in some cases. I believe this is due to floating point precision issues, so I've added a small tolerance in `onChangeResolution` to hopefully avoid this (I haven't been able to reproduce the problem with that in). Let me know if you've got a better idea for how to solve that!

PS. Thanks for spotting the `print` I left in for debugging in the previous commit